File: web/packages/drivetrain-sim/src/simulation/simulator.ts
/**
 * Drivetrain simulator with ODE integration.
 */

import { Drivetrain } from '../core/drivetrain';
import { interp } from '../math/interpolate';
import {
  SimulationConfig,
  DEFAULT_CONFIG,
  getOutputTimes,
} from './config';
import { SimulationResult, createResult } from './result';
import { solveIVP, integrateFixed } from './integrator';

/**
 * Control function type.
 * Takes time, state dict, and grade, returns control dict.
 */
export type ControlFunction = (
  t: number,
  state: Record<string, number>,
  grade: number
) => Record<string, number>;

/**
 * Grade function type.
 * Takes time, returns grade (fraction, e.g., 0.05 for 5%).
 */
export type GradeFunction = (t: number) => number;

/**
 * Drivetrain controller interface.
 */
export interface DrivetrainController {
  compute(state: Record<string, number>, grade: number): Record<string, number>;
}

/**
 * Simulator for any drivetrain topology.
 *
 * Wraps the Drivetrain dynamics with ODE integration and result processing.
 */
export class DrivetrainSimulator {
  readonly drivetrain: Drivetrain;
  private _controlLog: Array<[number, Record<string, number>]> = [];
  private _gradeLog: Array<[number, number]> = [];

  constructor(drivetrain: Drivetrain) {
    this.drivetrain = drivetrain;
  }

  /**
   * Run a time-domain simulation.
   *
   * @param x0 - Initial state {state_name: value}
   * @param controller - Control function or controller object
   * @param gradeProfile - Road grade as constant or function of time
   * @param config - Simulation configuration
   * @returns SimulationResult with all time series data
   */
  simulate(
    x0: Record<string, number>,
    controller: ControlFunction | DrivetrainController,
    gradeProfile: number | GradeFunction = 0.0,
    config: SimulationConfig = DEFAULT_CONFIG
  ): SimulationResult {
    this._controlLog = [];
    this._gradeLog = [];

    // Convert initial state to array
    const x0Arr = this.drivetrain.stateToArray(x0);

    // Wrap grade profile as function
    const gradeFn: GradeFunction =
      typeof gradeProfile === 'function'
        ? gradeProfile
        : () => gradeProfile as number;

    // Wrap controller
    const controlFn: ControlFunction =
      'compute' in controller
        ? (t, state, grade) => (controller as DrivetrainController).compute(state, grade)
        : (controller as ControlFunction);

    // Define dynamics wrapper for ODE solver
    const dynamicsWrapper = (t: number, x: number[]): number[] => {
      // Get current state as dict
      const state = this.drivetrain.arrayToState(x);

      // Get grade
      const grade = gradeFn(t);

      // Get control inputs
      const control = controlFn(t, state, grade);

      // Log for post-processing
      this._controlLog.push([t, { ...control }]);
      this._gradeLog.push([t, grade]);

      // Compute dynamics
      return this.drivetrain.dynamics(t, x, control, { grade });
    };

    // Set up output time points
    const tEval = getOutputTimes(config);

    // Run ODE integration with proper sub-stepping
    const sol = solveIVP(
      dynamicsWrapper,
      [config.tStart, config.tEnd],
      x0Arr,
      {
        method: config.method === 'RK45' ? 'RK45' : 'RK4',
        tEval,
        rtol: config.rtol,
        atol: config.atol,
        maxStep: config.maxStep ?? 0.005, // 5ms default for accuracy
      }
    );

    if (!sol.success) {
      throw new Error(`ODE integration failed: ${sol.message}`);
    }

    // Build result
    return this._buildResult(sol, config);
  }

  /**
   * Run simulation with progress callback (for UI updates).
   * Uses internal sub-stepping for accuracy (similar to scipy's adaptive stepping).
   */
  async simulateWithProgress(
    x0: Record<string, number>,
    controller: ControlFunction | DrivetrainController,
    gradeProfile: number | GradeFunction = 0.0,
    config: SimulationConfig = DEFAULT_CONFIG,
    onProgress?: (progress: number) => void
  ): Promise<SimulationResult> {
    this._controlLog = [];
    this._gradeLog = [];

    const x0Arr = this.drivetrain.stateToArray(x0);

    const gradeFn: GradeFunction =
      typeof gradeProfile === 'function'
        ? gradeProfile
        : () => gradeProfile as number;

    const controlFn: ControlFunction =
      'compute' in controller
        ? (t, state, grade) => (controller as DrivetrainController).compute(state, grade)
        : (controller as ControlFunction);

    const tEval = getOutputTimes(config);
    const totalSteps = tEval.length - 1;

    // Internal step size for accuracy (similar to scipy's adaptive stepping)
    const maxStep = config.maxStep ?? 0.005; // 5ms default for good accuracy

    // Manual integration with progress updates
    const yResult: number[][] = x0Arr.map(() => []);
    let x = [...x0Arr];
    let nfev = 0;

    // Record initial state
    for (let i = 0; i < x.length; i++) {
      yResult[i].push(x[i]);
    }

    // Helper for a single RK4 step
    const rk4Step = (t: number, state: number[], h: number, ctrl: Record<string, number>, grade: number): number[] => {
      const k1 = this.drivetrain.dynamics(t, state, ctrl, { grade });
      nfev++;
      const k2 = this.drivetrain.dynamics(
        t + h / 2,
        state.map((xi, i) => xi + (h / 2) * k1[i]),
        ctrl,
        { grade }
      );
      nfev++;
      const k3 = this.drivetrain.dynamics(
        t + h / 2,
        state.map((xi, i) => xi + (h / 2) * k2[i]),
        ctrl,
        { grade }
      );
      nfev++;
      const k4 = this.drivetrain.dynamics(
        t + h,
        state.map((xi, i) => xi + h * k3[i]),
        ctrl,
        { grade }
      );
      nfev++;

      return state.map(
        (xi, i) => xi + (h / 6) * (k1[i] + 2 * k2[i] + 2 * k3[i] + k4[i])
      );
    };

    for (let k = 1; k < tEval.length; k++) {
      const tStart = tEval[k - 1];
      const tEnd = tEval[k];

      // Sub-step through this interval for accuracy
      let t = tStart;
      while (t < tEnd - 1e-12) {
        // Determine step size (don't overshoot)
        const h = Math.min(maxStep, tEnd - t);

        // Get current state dict and control at this sub-step time
        const stateDict = this.drivetrain.arrayToState(x);
        const grade = gradeFn(t);
        const control = controlFn(t, stateDict, grade);

        // Only log at output interval boundaries (not every sub-step)
        if (Math.abs(t - tStart) < 1e-12) {
          this._controlLog.push([t, { ...control }]);
          this._gradeLog.push([t, grade]);
        }

        // RK4 step
        x = rk4Step(t, x, h, control, grade);
        t += h;
      }

      // Record state at output time
      for (let i = 0; i < x.length; i++) {
        yResult[i].push(x[i]);
      }

      // Report progress
      if (onProgress && k % 10 === 0) {
        onProgress(k / totalSteps);
        // Yield to allow UI updates
        await new Promise((resolve) => setTimeout(resolve, 0));
      }
    }

    if (onProgress) {
      onProgress(1);
    }

    const sol = {
      t: tEval,
      y: yResult,
      success: true,
      nfev,
    };

    return this._buildResult(sol, config);
  }

  private _buildResult(
    sol: { t: number[]; y: number[][]; nfev: number },
    config: SimulationConfig
  ): SimulationResult {
    const time = sol.t;
    const nPoints = time.length;

    // Extract states
    const states: Record<string, number[]> = {};
    for (let i = 0; i < this.drivetrain.stateNames.length; i++) {
      states[this.drivetrain.stateNames[i]] = sol.y[i];
    }

    // Interpolate control log to output times
    const controls = this._interpolateControls(time);

    // Compute derived outputs
    const outputs = this._computeOutputs(time, sol.y, controls);

    // Metadata
    const metadata = {
      solver_method: config.method,
      n_function_evals: sol.nfev,
      drivetrain_type: 'Drivetrain',
      components: Array.from(this.drivetrain.topology.components.keys()),
    };

    return {
      time,
      states,
      controls,
      outputs,
      metadata,
    };
  }

  private _interpolateControls(time: number[]): Record<string, number[]> {
    if (this._controlLog.length === 0) {
      return {};
    }

    // Get all control keys
    const allKeys = new Set<string>();
    for (const [, ctrl] of this._controlLog) {
      for (const key of Object.keys(ctrl)) {
        allKeys.add(key);
      }
    }

    // Build time series for each control
    const logTimes = this._controlLog.map(([t]) => t);
    const controls: Record<string, number[]> = {};

    for (const key of allKeys) {
      const values = this._controlLog.map(([, ctrl]) => ctrl[key] ?? 0);
      controls[key] = time.map((t) => interp(t, logTimes, values));
    }

    return controls;
  }

  private _computeOutputs(
    time: number[],
    y: number[][],
    controls: Record<string, number[]>
  ): Record<string, number[]> {
    const outputs: Record<string, number[]> = {};
    const nPoints = time.length;

    // Velocity
    const velocities: number[] = [];
    for (let i = 0; i < nPoints; i++) {
      const state = y.map((arr) => arr[i]);
      velocities.push(this.drivetrain.getVelocity(state));
    }
    outputs.velocity = velocities;

    // Grade (from log)
    if (this._gradeLog.length > 0) {
      const logTimes = this._gradeLog.map(([t]) => t);
      const grades = this._gradeLog.map(([, g]) => g);
      outputs.grade = time.map((t) => interp(t, logTimes, grades));
    }

    // Power calculations
    Object.assign(outputs, this._computePowerOutputs(time, y, controls));

    return outputs;
  }

  private _computePowerOutputs(
    time: number[],
    y: number[][],
    controls: Record<string, number[]>
  ): Record<string, number[]> {
    const outputs: Record<string, number[]> = {};
    const nPoints = time.length;
    const components = this.drivetrain.topology.components;

    // Engine power
    for (const [compName, component] of components) {
      if (component.constructor.name.includes('Engine')) {
        const tKey = `T_${compName}`;
        if (tKey in controls) {
          const speedKey = `${compName}.shaft`;
          const omega: number[] = [];

          for (let i = 0; i < nPoints; i++) {
            const state = y.map((arr) => arr[i]);
            const allSpeeds = this.drivetrain.getAllSpeeds(state);
            omega.push(allSpeeds[speedKey] ?? 0);
          }

          const torque = controls[tKey];
          const power = torque.map((t, i) => t * omega[i]);
          outputs[`P_${compName}`] = power;

          // Fuel rate
          if ('getFuelRate' in component) {
            const fuelRate: number[] = [];
            for (let i = 0; i < nPoints; i++) {
              fuelRate.push((component as any).getFuelRate(torque[i], omega[i]));
            }
            outputs.fuel_rate = fuelRate;
          }
        }
      }
    }

    // Motor power
    for (const [compName, component] of components) {
      if (component.constructor.name.includes('Motor')) {
        const tKey = `T_${compName}`;
        if (tKey in controls) {
          const speedKey = `${compName}.shaft`;
          const omega: number[] = [];

          for (let i = 0; i < nPoints; i++) {
            const state = y.map((arr) => arr[i]);
            const allSpeeds = this.drivetrain.getAllSpeeds(state);
            omega.push(allSpeeds[speedKey] ?? 0);
          }

          const torque = controls[tKey];
          const pMech = torque.map((t, i) => t * omega[i]);
          outputs[`P_${compName}_mech`] = pMech;

          // Electrical power
          if ('getElectricalPower' in component) {
            const pElec: number[] = [];
            for (let i = 0; i < nPoints; i++) {
              pElec.push((component as any).getElectricalPower(torque[i], omega[i]));
            }
            outputs[`P_${compName}_elec`] = pElec;
          }
        }
      }
    }

    return outputs;
  }
}

/**
 * Convenience function to run a simulation.
 */
export function simulate(
  drivetrain: Drivetrain,
  x0: Record<string, number>,
  controller: ControlFunction | DrivetrainController,
  gradeProfile: number | GradeFunction = 0.0,
  config: SimulationConfig = DEFAULT_CONFIG
): SimulationResult {
  const simulator = new DrivetrainSimulator(drivetrain);
  return simulator.simulate(x0, controller, gradeProfile, config);
}

File: src/gearbox_sim/simulation/simulator.py
"""Drivetrain simulator with ODE integration."""

from typing import Callable, Dict, List, Optional, Union

import numpy as np
from numpy.typing import NDArray
from scipy.integrate import solve_ivp

from ..core.drivetrain import Drivetrain
from .config import SimulationConfig
from .result import SimulationResult


# Type aliases
ControlFunction = Callable[[float, Dict[str, float], float], Dict[str, float]]
GradeFunction = Callable[[float], float]


class DrivetrainSimulator:
    """Simulator for any drivetrain topology.

    Wraps the Drivetrain dynamics with ODE integration and result processing.
    """

    def __init__(self, drivetrain: Drivetrain):
        """Initialize the simulator.

        Args:
            drivetrain: Compiled drivetrain to simulate
        """
        self.drivetrain = drivetrain
        self._control_log: List[tuple] = []
        self._grade_log: List[tuple] = []

    def simulate(
        self,
        x0: Dict[str, float],
        controller: Union[ControlFunction, "DrivetrainController"],
        grade_profile: Union[float, GradeFunction] = 0.0,
        config: SimulationConfig = None,
    ) -> SimulationResult:
        """Run a time-domain simulation.

        Args:
            x0: Initial state {state_name: value}
            controller: Control function or controller object
                Function signature: (t, state_dict, grade) -> control_dict
            grade_profile: Road grade as constant or function of time
                Function signature: (t) -> grade
            config: Simulation configuration

        Returns:
            SimulationResult with all time series data
        """
        config = config or SimulationConfig()
        self._control_log = []
        self._grade_log = []

        # Convert initial state to array
        x0_arr = self.drivetrain.state_to_array(x0)

        # Wrap grade profile as function
        if callable(grade_profile):
            grade_fn = grade_profile
        else:
            grade_fn = lambda t: float(grade_profile)

        # Wrap controller
        if hasattr(controller, "compute"):
            control_fn = lambda t, state, grade: controller.compute(state, grade)
        else:
            control_fn = controller

        # Define dynamics wrapper for ODE solver
        def dynamics_wrapper(t: float, x: NDArray) -> NDArray:
            # Get current state as dict
            state = self.drivetrain.array_to_state(x)

            # Get grade
            grade = grade_fn(t)

            # Get control inputs
            control = control_fn(t, state, grade)

            # Log for post-processing
            self._control_log.append((t, control.copy()))
            self._grade_log.append((t, grade))

            # Compute dynamics
            return self.drivetrain.dynamics(t, x, control, {"grade": grade})

        # Set up output time points
        t_eval = config.get_output_times()

        # Run ODE integration
        sol = solve_ivp(
            dynamics_wrapper,
            (config.t_start, config.t_end),
            x0_arr,
            method=config.method,
            t_eval=t_eval,
            rtol=config.rtol,
            atol=config.atol,
            max_step=config.max_step if config.max_step else np.inf,
        )

        if not sol.success:
            raise RuntimeError(f"ODE integration failed: {sol.message}")

        # Build result
        return self._build_result(sol, config)

    def _build_result(self, sol, config: SimulationConfig) -> SimulationResult:
        """Build SimulationResult from ODE solution."""
        time = sol.t
        n_points = len(time)

        # Extract states
        states = {}
        for i, name in enumerate(self.drivetrain.state_names):
            states[name] = sol.y[i, :]

        # Interpolate control log to output times
        controls = self._interpolate_controls(time)

        # Compute derived outputs
        outputs = self._compute_outputs(time, sol.y, controls)

        # Metadata
        metadata = {
            "solver_method": config.method,
            "n_function_evals": sol.nfev,
            "drivetrain_type": type(self.drivetrain).__name__,
            "components": list(self.drivetrain.topology.components.keys()),
        }

        return SimulationResult(
            time=time,
            states=states,
            controls=controls,
            outputs=outputs,
            metadata=metadata,
        )

    def _interpolate_controls(self, time: NDArray) -> Dict[str, NDArray]:
        """Interpolate logged control inputs to output times."""
        if not self._control_log:
            return {}

        # Get all control keys
        all_keys = set()
        for _, ctrl in self._control_log:
            all_keys.update(ctrl.keys())

        # Build time series for each control
        log_times = np.array([t for t, _ in self._control_log])
        controls = {}

        for key in all_keys:
            values = np.array([ctrl.get(key, 0.0) for _, ctrl in self._control_log])
            controls[key] = np.interp(time, log_times, values)

        return controls

    def _compute_outputs(
        self, time: NDArray, y: NDArray, controls: Dict[str, NDArray]
    ) -> Dict[str, NDArray]:
        """Compute derived output quantities."""
        outputs = {}
        n_points = len(time)

        # Velocity
        velocities = np.zeros(n_points)
        for i in range(n_points):
            velocities[i] = self.drivetrain.get_velocity(y[:, i])
        outputs["velocity"] = velocities

        # Grade (from log)
        if self._grade_log:
            log_times = np.array([t for t, _ in self._grade_log])
            grades = np.array([g for _, g in self._grade_log])
            outputs["grade"] = np.interp(time, log_times, grades)

        # Power calculations
        outputs.update(self._compute_power_outputs(time, y, controls))

        return outputs

    def _compute_power_outputs(
        self, time: NDArray, y: NDArray, controls: Dict[str, NDArray]
    ) -> Dict[str, NDArray]:
        """Compute power-related outputs."""
        outputs = {}
        n_points = len(time)
        components = self.drivetrain.topology.components

        # Engine power
        for comp_name, component in components.items():
            if "Engine" in type(component).__name__:
                T_key = f"T_{comp_name}"
                if T_key in controls:
                    # Find engine speed in states
                    speed_key = f"{comp_name}.shaft"
                    omega = np.zeros(n_points)
                    speeds = self.drivetrain.get_all_speeds(y[:, 0])

                    for i in range(n_points):
                        all_speeds = self.drivetrain.get_all_speeds(y[:, i])
                        omega[i] = all_speeds.get(speed_key, 0.0)

                    torque = controls[T_key]
                    power = torque * omega
                    outputs[f"P_{comp_name}"] = power

                    # Fuel rate
                    if hasattr(component, "get_fuel_rate"):
                        fuel_rate = np.zeros(n_points)
                        for i in range(n_points):
                            fuel_rate[i] = component.get_fuel_rate(torque[i], omega[i])
                        outputs["fuel_rate"] = fuel_rate

        # Motor power
        for comp_name, component in components.items():
            if "Motor" in type(component).__name__:
                T_key = f"T_{comp_name}"
                if T_key in controls:
                    speed_key = f"{comp_name}.shaft"

                    omega = np.zeros(n_points)
                    for i in range(n_points):
                        all_speeds = self.drivetrain.get_all_speeds(y[:, i])
                        omega[i] = all_speeds.get(speed_key, 0.0)

                    torque = controls[T_key]
                    P_mech = torque * omega
                    outputs[f"P_{comp_name}_mech"] = P_mech

                    # Electrical power
                    if hasattr(component, "get_electrical_power"):
                        P_elec = np.zeros(n_points)
                        for i in range(n_points):
                            P_elec[i] = component.get_electrical_power(torque[i], omega[i])
                        outputs[f"P_{comp_name}_elec"] = P_elec

        return outputs


def simulate(
    drivetrain: Drivetrain,
    x0: Dict[str, float],
    controller: Union[ControlFunction, "DrivetrainController"],
    grade_profile: Union[float, GradeFunction] = 0.0,
    config: SimulationConfig = None,
) -> SimulationResult:
    """Convenience function to run a simulation.

    Args:
        drivetrain: Compiled drivetrain
        x0: Initial state
        controller: Control function or controller object
        grade_profile: Road grade (constant or function)
        config: Simulation configuration

    Returns:
        SimulationResult
    """
    simulator = DrivetrainSimulator(drivetrain)
    return simulator.simulate(x0, controller, grade_profile, config)
